### ISSUE TITLE
Vector3.Normalize(), .Distance(), and .Equals() do not return correct values in some cases

### ISSUE THREAD

**Issue by mellinoe** [dotnet/coreclr#8220]
I've been trying to track down an obscure issue with a project I am working on, and I believe I've found the root cause to be related to `Vector3.Normalize` not returning correct values. I haven't yet been able to get a very minimal repro, but the problem always occurs in the context of my project (which is fairly compex). The scenario is as follows:

* Running an anonymous delegate with one parameter, `ref ParticleState`, where `ParticleState` has a `Vector3 Offset` field (among other stuff).
* Subtract `ParticleState.Offset` from another `Vector3`, retrieved from some other object.
* Call `Vector3.Normalize` on the result of the subtraction.
* Use that normalized direction vector later in the function.

The actual code can be found here: https://github.com/mellinoe/ge-test-game/blob/0939e2496a1cca1229d504643055a4de3c1ce932/code/src/GravityGame/ParticleVacuumEffect.cs#L49

I did some experiments and have a branch [here](https://github.com/mellinoe/ge-test-game/blob/simd-repro/code/src/GravityGame/ParticleVacuumEffect.cs#L49) that prints out info about the problem. The gist of it is that manually computing the normalized difference vector gives a different result than using `Vector3.Normalize(a - b)`. Also, the `.Length()` of values returned from `Vector3.Normalize()` are not always equal to 1, which should be impossible.

```CSharp
// SIMD BUG
Vector3 directionBad = Vector3.Normalize(_vacuumTarget.Position - ps.Offset); // This should be equivalent to "direction" on the next two lines.
Vector3 direction = new Vector3(_vacuumTarget.Position.X - ps.Offset.X, _vacuumTarget.Position.Y - ps.Offset.Y, _vacuumTarget.Position.Z - ps.Offset.Z);
direction = direction / direction.Length();
```
`directionBad != direction`, and `directionBad.Length() != 1`.

Setting `COMPLUS_FEATURESIMD=0` causes the problem to no longer repro. When that is set, the two values above (`directionBad` and `direction`) are identical in all cases.

The issue is also very context-sensitive. When I assign some intermediate results to variables, it no longer repros. For example, if I assign the `_vacuumTarget.Position - ps.Offset` result to a variable, and call `Vector3.Normalize` on that, then the problem doesn't repro. I also use `Vector3.Normalize` in a ton of other places in this project and have not seen issues like this. I suspect it might have something to do with the ref parameter to a struct, or something like that.

I can help investigate and get a repro set up if needed, since I'm not able to find a simpler repro right now. I've been able to repro this on both the 1.0 and 1.1 CoreCLR runtimes, on Windows. I haven't seen if this repros on other OS's, although I could try to check that on Monday.

@CarolEidt , @sivarv

**Comment by mellinoe** [comment]
Another important thing: I didn't notice this for a very long time because it does not repro in Debug builds. It only repros in Release mode builds.

**Comment by mellinoe** [comment]
I tracked down a couple more places that were behaving differently in Debug and Release builds, or with `COMPLUS_FEATURESIMD=0`. These commits were what I needed to do to work around the problem:

https://github.com/mellinoe/ge-test-game/commit/78f682085d5214a62d7a353c05bce99a994cd2d6
https://github.com/mellinoe/ge-test-game/commit/98405dd6e27d62acbfd44b80a46086eef4286860

The fixes were essentially manually hoisting some references into local variables, and then doing operations on those variables. Everything else is otherwise identical. These do all seem to be related to SIMD codegen, because the problems all go away when `COMPLUS_FEATURESIMD` is set to 0.
